File: func.rb
Paper = Struct.new(
  :title,          # paper title (String)
  :id,             # paper id (Integer)
  :author,         # list of author (Author Struct)
  :citation_count, # number of times cited (Integer)
  :abstract,       # abstract (String)
  :conference,     # conference (Conference Struct)
  :journal,        # journal (Journal Struct)
  :reference,      # list of reference
  :tier
)
Author = Struct.new(
  :name,
  :affiliation
)
Conference = Struct.new(
  :id,
  :name
)
Journal = Struct.new(
  :id,
  :name
)
Publication = Struct.new(
  :type,
  :name_full,
  :name_short,
  :publisher,
  :volume,
  :year
)

def read_attributes(_path)

  attributes = ''
  File.foreach(_path) { |l|
    next if l.start_with?('//')
    next if l.length == 0
    attributes << "#{l.chomp},"
  }

  return attributes.chop

end

def abstract(_ia)

  return "" if _ia.nil?

  tmp = Array.new(_ia['IndexLength'])

  _ia['InvertedIndex'].each { |k,v|
    v.each { |i| tmp[i] = k }
  }

  return tmp.join(' ')

end

#
def parse_entity(_e, _p)

  _p['title'] = _e['DN']

  _p['id'] = _e['Id']

  num_author = _e['AA'].length
  authors = Array.new(num_author)
  _e['AA'].each { |a|
    index = a['S'] - 1 # index start from 1
    authors[index] = Author.new(a['DAuN'], a['DAfN'])
  }
  _p['author'] = authors

  _p['citation_count'] = _e['CC']

  _p['abstract'] = abstract(_e['IA'])

  c = _e['C']
  _p['conference'] = c.nil? ? Conference.new : Conference.new(c['CId'], c['CN']) 

  j = _e['J']
  _p['journal'] = j.nil? ? Journal.new : Journal.new(j['JId'], j['JN'])

  _p['reference'] = _e['RId']

end

def evaluate(_json, _papers, _forest, _tier)

  entities = _json['entities']

  entities.each_with_index { |e,i|
    # ----------------------
    # process for _papers
    # ----------------------
    if _papers.has_key?(e['Id'])
      # add tier info if _papers has ID
      _papers[e['Id']]['tier'] << _tier
    else
      # create paper if _papers does not have ID
      p = Paper.new
      p['tier'] = Array.new
      p['tier'] << _tier
      parse_entity(e, p)
      _papers[e['Id']] = p
    end
    # ----------------------
    # process for _forest
    # ----------------------
    e['RId'].each { |r| 
      _forest[e['Id']] = Array.new unless _forest.has_key?(e['Id'])
      _forest[e['Id']] << r
    } if e.has_key?('RId')
  }

end

File: src/func.py
import requests

def invoke_api(url, headers):

    response = requests.get(
        url,
        headers = headers
    )

    return response

def invoke_evaluate(token, expr, attr, count = 10, offset = 0, model = 'latest'):
    options = ''
    if expr  is not None: options += f'expr={expr}&'
    if attr  is not None: options += f'attributes={attr}&'
    options += f'model={model}&'
    options += f'count={count}&'
    options += f'offset={offset}&'
    options = options[:-1]

    url = f'https://api.labs.cognitive.microsoft.com/academic/v1.0/evaluate?{options}'
    headers = {
        'Content-Type' : 'application/json',
        'Ocp-Apim-Subscription-Key' : token
    }

    print(f"url: {url}")
    response = invoke_api(url, headers)

    return response.json()

def parse_abstract(info):

    l = info['IndexLength']
    abst = [''] * l

    tmp = info['InvertedIndex']
    for k in tmp:
        for v in tmp[k]:
            abst[v] = k
    
    return ' '.join(abst)

def parse_authors(authors):
    l = len(authors)
    retval = [{} for i in range(l)]
    for i,a in enumerate(authors):
        r = retval[i]
        r['name'] = a['DAuN']
        r['affiliation'] = a['DAfN']
    return retval

def pub_type_name(pt):
    return ['Unknown',
        'Journal',
        'Patent',
        'Conference',
        'Book',
        'Book',
        'Book',
        'Dataset',
        'Repository'][pt]

def parse_entities(entities, papers):
    count = 0
    for e in entities:
        id = str(e['Id'])
        if id in papers:
            continue
        # papers
        p = {}
        p['abst'] = parse_abstract(e['IA']) if 'IA' in e else 'Unknown'
        p['authors'] = parse_authors(e['AA']) if 'AA' in e else 'Unknown'
        p['citation_count'] = int(e['CC']) if 'CC' in e else 'Unknown'
        p['conference'] = {'id':e['C']['CId'],'name':e['C']['CN']} if 'C' in e else 'Unknown'
        p['id'] = id if 'Id' in e else 'Unknown'
        p['journal-id'] = e['J']['JId'] if 'J' in e and 'JId' in e else 'Unknown'
        p['journal-name'] = e['J']['JN'] if 'J' in e and 'JN' in e else 'Unknown'
        p['pub-type'] = e['Pt'] if 'Pt' in e else 'Unknown'
        p['pub-name_f'] = e['VFN'] if 'VFN' in e else 'Unknown'
        p['pub-name_s'] = e['VSN'] if 'VSN' in e else 'Unknown'
        p['publisher'] = e['PB'] if 'PB' in e else 'Unknown'
        p['volume'] = int(e['V']) if 'V' in e else 'Unknown'
        p['year'] = int(e['Y']) if 'Y' in e else '1000'
        p['title'] = e['DN'] if 'DN' in e else 'Unknown'
        p['references'] = list(map(lambda x: str(x), e['RId'])) if 'RId' in e else []
        p['citcon'] = e['CitCon'] if 'CitCon' in e else 'Unknown'
        papers[id] = p
        count += 1
    return count
